observer/holder/nodesHolder.go

```diff
@@ -0,0 +1,217 @@
+		shouldSkipNode := nh.availability == data.AvailabilityRecent && !node.IsSnapshotless ||
+			nh.availability == data.AvailabilityAll && node.IsSnapshotless
+		if shouldSkipNode {
+			continue
+		}
+		nh.allNodes[node.ShardId] = append(nh.allNodes[node.ShardId], node)
+	}
+
+	nh.printNodesInShardsUnprotected()
+}
+
+// GetSyncedNodes returns all the synced nodes
+func (nh *nodesHolder) GetSyncedNodes(shardID uint32) []*data.NodeData {
+	return nh.getObservers(syncedNodesCache, shardID)
+}
+
+// GetSyncedFallbackNodes returns all the synced fallback nodes
+func (nh *nodesHolder) GetSyncedFallbackNodes(shardID uint32) []*data.NodeData {
+	return nh.getObservers(syncedFallbackCache, shardID)
+}
+
+// GetOutOfSyncNodes returns all the out of sync nodes
+func (nh *nodesHolder) GetOutOfSyncNodes(shardID uint32) []*data.NodeData {
+	return nh.getObservers(outOfSyncNodesCache, shardID)
+}
+
+// GetOutOfSyncFallbackNodes returns all the out of sync fallback nodes
+func (nh *nodesHolder) GetOutOfSyncFallbackNodes(shardID uint32) []*data.NodeData {
+	return nh.getObservers(outOfSyncFallbackNodes, shardID)
+}
+
+// Count computes and returns the total number of nodes
+func (nh *nodesHolder) Count() int {
+	counter := 0
+	nh.mut.RLock()
+	defer nh.mut.RUnlock()
+
+	for _, nodes := range nh.allNodes {
+		counter += len(nodes)
+	}
+
+	return counter
+}
+
+func (nh *nodesHolder) getObservers(cache cacheType, shardID uint32) []*data.NodeData {
+	cacheKey := getCacheKey(cache, shardID)
+	nh.mut.RLock()
+	cachedValues, exists := nh.cache[getCacheKey(cache, shardID)]
+	nh.mut.RUnlock()
+
+	if exists {
+		return cachedValues
+	}
+
+	// nodes not cached, compute the list and update the cache
+	recomputedList := make([]*data.NodeData, 0)
+	nh.mut.Lock()
+	for _, node := range nh.allNodes[shardID] {
+		if areCompatibleParameters(cache, node) {
+			recomputedList = append(recomputedList, node)
+		}
+	}
+	nh.cache[cacheKey] = recomputedList
+	nh.mut.Unlock()
+
+	return recomputedList
+}
+
+func areCompatibleParameters(cache cacheType, node *data.NodeData) bool {
+	isSynced, isFallback := node.IsSynced, node.IsFallback
+	if cache == syncedFallbackCache && isSynced && isFallback {
+		return true
+	}
+	if cache == outOfSyncFallbackNodes && !isSynced && isFallback {
+		return true
+	}
+	if cache == syncedNodesCache && isSynced && !isFallback {
+		return true
+	}
+	if cache == outOfSyncNodesCache && !isSynced && !isFallback {
+		return true
+	}
+
+	return false
+}
+
+func getCacheKey(cache cacheType, shardID uint32) string {
+	return fmt.Sprintf("%s_%d", cache, shardID)
+}
+
+// IsInterfaceNil returns true if there is no value under the interface
+func (nh *nodesHolder) IsInterfaceNil() bool {
+	return nh == nil
+}
+
+func (nh *nodesHolder) printNodesInShardsUnprotected() {
+	nodesByType := make(map[uint32]map[cacheType][]*data.NodeData)
+
+	// define a function to get the cache type for a node
+	getCacheType := func(node *data.NodeData) cacheType {
```

Comment:
why not a proper function? Anonymous functions have their cost in terms of performance

Reply:
extracted a new function